lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+-- table has named fields. It first checks for red, green, blue fields (optional
+-- alpha field); then for hue, saturation, lightness; then if it has 3 or 4
+-- fields. If it gets to a 3 or 4 field table, it will check the first for HSL
+-- values, then for RGB.
+--
+-- After a color is created you can access it's individual elements as needed.
+--     c = colorful.color.new("#ffcc99")
+--     c.red        -- 1.0
+--     c.green      -- 0.8
+--     c.blue       -- 0.6
+--     c.hue        -- 30
+--     c.saturation -- 80%
+--     c.lightness  -- 60%
+--     c.alpha      -- 1.0 Note this defaults to no transparency
+--
+-- Once a color is created you can do transformations on it. Note that they do
+-- not modify the original color object, but instead return a new one.
+--     c2 = colorful.color("#99ccff")
+--     c3 = c:mix(c2, 0.4) -- mix c with c2, giving c 40% and c2 60%
+--     c:darken(0.25) -- darken 25%
+--     c:lighten(0.25) -- lighten 25%
+--     c:saturate(0.1) -- saturate 10%
+--     c:desaturate(0.1) -- desaturate 10%
+--     c:grayscale() -- desaturate color completely to grayscale
+--     c:invert() -- inverts color
+--     c:complement() -- the compliment of the color (180deg hue shift)
+--
+-- If you want to create a color scheme you can use the color harmony functions:
+--     local left, right = c:analogous()
+--     local left, right = c:split_complementary()
+--     local left, right = c:triadic()
+--     local complement, analogous, analogous_complement = c:tetradic()
+--
+-- @author Kevin Zander
+-- @copyright 2019 Kevin Zander
+-- @module beautiful.colorful
+---------------------------------------------------------------------------
+
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+local colorful = {}
+
+local color = {
+    -- set the from and as tables
+    from = {},
+    as = {}
+}
+
+-- Metatable for color instances. This is set up near the end of the file.
+local color_mt = {}
+
+-- {{{ Helper functions
+
+-- Lua 5.1 does not support bit shifting operators
+local function lshift(x, disp)
+    return (x * 2^disp) % 2^32
+end
+
+local function rshift(x, disp)
+    return math.floor(x % 2^32 / 2^disp)
+end
+
+-- Do we want to pull in gears.math.round?
+local function round(x)
+    local d,_ = math.modf(x + (x >= 0 and 1 or -1) * 0.5)
+    return d
+end
+
+-- Clamp a number between a low and high
+local function clamp(v, l, h)
+    if v < l then return l end
+    if v > h then return h end
+    return v
+end
+
+-- Return true/false if a number is between a low and high
+local function between(v, l, h)
+    if l <= v and v <= h then return true end
+    return false
+end
+
+-- Expects Hue [0,360], Saturation [0,1], Lightness [0,1]
+-- Does NOT check bounds
+local function hsl_to_rgb(hue, saturation, lightness)
+    -- If s = 0 then it's a grey
+    if saturation == 0 then
+        return unpack({lightness, lightness, lightness})
```

Comment:
(Same comment applies to all other uses of `unpack` in this file)
Why is it done like this?
The above is equivalent to `return lightness, lightness, lightness`.

Reply:
I thought I remembered seeing it used in other places in the API, I could be mistaken. Not a problem to change it.